inputremapper/gui/components/gtkext/listbox_filter.py

```diff
@@ -0,0 +1,133 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with input-remapper.  If not, see <https://www.gnu.org/licenses/>.
+
+import gi
+
+from gi.repository import Gtk
+
+from typing import Iterator
+
+
+class ListBoxFilter:
+    """Implements UI-side filtering of list widgets.
+
+    The following example creates a new ``ListBoxFilter`` for a given ``Gtk.ListBox``.
+    It also sets all optional arguments to override some default behavior.
+
+    >>> filter = ListBoxFilter(
+    >>>     my_listbox,                  # Gtk.ListBox to be managed
+    >>>     get_row_name=MyRow.get_name  # custom row name getter
+    >>>     filter_value="text"          # inital value
+    >>>     case_sensitive=True,         # override default:False
+    >>> )
+
+    To apply a filter use `set_filter` as follows.
+
+    >>> filter.set_filter("some text")
+    >>> filter.set_filter("More Text", case_sensitive=True)
+
+    """
+
+    MAX_WIDGET_TREE_TEXT_SEARCH_DEPTH = 10
+
+    def __init__(
+        self,
+        listbox: Gtk.ListBox,
+        get_row_name=None,
+        filter_value="",
+        case_sensitive=False,
+    ):
+        self._controlled_listbox: Gtk.ListBox = listbox
+        self._get_row_name = get_row_name or self.get_row_name
+        self._filter_value: str = ""
+        self._case_sensitive = False
+        self.set_filter(filter_value, case_sensitive=case_sensitive)
+
+    @classmethod
+    def get_row_name(T, row: Gtk.ListBoxRow) -> str:
+        """
+        Returns the visible text of a Gtk.ListBoxRow from both the row's `name`
+        attribute or the row's text in the UI.
+        """
+        text = getattr(row, "name", "")
+
+        # find and join all text in the ListBoxRow
+        text += " ".join(v for v in T.get_widget_tree_text(row) if v != "")
+
+        return text.strip()
+
+    @classmethod
+    def get_widget_tree_text(T, widget: Gtk.Widget, level=0) -> Iterator[str]:
+        """
+        Recursively traverses the tree of child widgets starting from the given
+        widget, and yields the text of all text-containing widgets.
+        """
+        if level > T.MAX_WIDGET_TREE_TEXT_SEARCH_DEPTH:
+            return
+
+        if hasattr(widget, "get_label"):
+            yield (widget.get_label() or "").strip()
+        if hasattr(widget, "get_text"):
+            yield (widget.get_text() or "").strip()
+        if isinstance(widget, Gtk.Container):
+            for t in widget.get_children():
+                yield from T.get_widget_tree_text(t, level=level + 1)
+
+    @property
+    def filter_value(self):
+        return self._filter_value
+
+    @property
+    def case_sensitive(self):
+        return self._case_sensitive
+
+    # matches the current filter_value and filter_options with the given value
+    def match_filter(self, value: str):
```

Comment:
those should probably be docstrings instead

```suggestion
    def match_filter(self, value: str):
        """Match the current filter_value and filter_options with the given value."""
```

here and everywhere else